test/soaked_test.js

```diff
@@ -1,143 +1,348 @@
 
-describe.skip('soaked expressions', () => {
+describe('soaked expressions', () => {
   describe('function application', () => {
     it('works with a basic function', () => {
       check(`
         a?()
       `, `
-        if (typeof a === 'function') { a(); }
+        __guardFunc__(a, f => f());
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
 
     it('works with a function that is not safe to repeat', () => {
       check(`
         a()?()
       `, `
-        let fn;
-        if (typeof (fn = a()) === 'function') { fn(); }
+        __guardFunc__(a(), f => f());
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
 
     it('works in an expression context', () => {
       check(`
         a(b()?())
       `, `
-        let fn;
-        a(typeof (fn = b()) === 'function' ? fn() : undefined);
+        a(__guardFunc__(b(), f => f()));
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
 
     it('preserves arguments', () => {
         check(`
         a?(1, 2, 3)
       `, `
-        if (typeof a === 'function') { a(1, 2, 3); }
+        __guardFunc__(a, f => f(1, 2, 3));
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
 
-    it('combines conditions of nested soaked function calls', () => {
+    it('handles nested soaked function calls', () => {
       check(`
         a?(1)?(2)
       `, `
-        let fn;
-        if (typeof a === 'function' && typeof (ref = a(1)) === 'function') { fn(2); }
+        __guardFunc__(__guardFunc__(a, f1 => f1(1)), f => f(2));
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
 
     it('works with dynamic member access as the function', () => {
       check(`
         a[b]?()
       `, `
-        if (typeof a[b] === 'function') { a[b](); }
+        __guardFunc__(a[b], f => f());
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
 
     it('works with dynamic member access whose key is unsafe to repeat as the function', () => {
       check(`
         a[b()]?()
       `, `
-        let fn;
-        if (typeof (fn = a[b()]) === 'function') { fn(); }
+        __guardFunc__(a[b()], f => f());
+        function __guardFunc__(func, transform) {
+          return typeof func === 'function' ? transform(func) : undefined;
+        }
       `);
     });
+
+    it('evaluates soaked function calls', () => {
+      validate(`
+        f = -> 3
+        o = f?()
+      `, 3);
```

Comment:
Note that I'm using the change to `validate` from #321 here. Without that change, the `3` argument is ignored (although it's still useful documentation, in my opinion), but once #321 lands, `validate` will also make sure that the result is `3` (not just that CoffeeScript and Decaffeinate produce the same result).